`subTotalcost` panics when trying to delete expired transaction in `encrypted` queue

#### System information

Geth version: `v0.3.1`
OS & Version: Linux

#### Expected behaviour

Any transactions in Geth queue (as well as the `encrypted` cache after #316) has a 3h-lifetime limit as default. After being added into mempool and get expired, these transactions should be removed safely.

#### Actual behaviour

Currently, on T4, we've set up such a node with `--txpool.signaturecache` enabled, and it panicked with `totalcost underflow` when `removeTx`. BTW, this happened when T4 stopped, which means no block was generated during that 3 hours.

#### Steps to reproduce the behaviour


#### Backtrace

````
INFO [03-20|07:56:24.943] encrypted transaction cached             txHash=ef0796..46dc97 sender=0x2e211B562507169E1C444bA39d77feD711a69A89 nonce=162
WARN [03-20|07:56:24.943] Served eth_sendRawTransaction            conn=127.0.0.1:44542 reqid=1579251176812223 duration=1.220517ms err="encrypted transaction cached"
WARN [03-20|07:56:37.833] Served eth_call                          conn=127.0.0.1:57634 reqid=0c755813-ea2e-4024-a4f4-84c164d5d37a duration="466.739µs" err="execution reverted" errdata=0x7352d91c
WARN [03-20|07:56:38.030] Served eth_call                          conn=127.0.0.1:57638 reqid=01d60508-5fa9-4cd8-b293-b745cb55b6f2 duration="450.358µs" err="execution reverted" errdata=0x7352d91c
WARN [03-20|07:56:38.237] Served eth_call                          conn=127.0.0.1:57648 reqid=d0d0eeb0-f58c-45e9-bc63-a4058673249f duration="447.251µs" err="execution reverted" errdata=0x7352d91c
WARN [03-20|07:56:38.432] Served eth_call                          conn=127.0.0.1:57662 reqid=2c241b8f-d87a-4ebf-ad01-1e1eef30f735 duration="507.322µs" err="execution reverted" errdata=0x7352d91c
INFO [03-20|08:48:35.799] Regenerated local transaction journal    transactions=2 accounts=1
INFO [03-20|09:48:35.799] Regenerated local transaction journal    transactions=2 accounts=1
INFO [03-20|10:48:35.799] Regenerated local transaction journal    transactions=2 accounts=1
panic: totalcost underflow

goroutine 4192 [running]:
github.com/ethereum/go-ethereum/core/txpool/legacypool.(*list).subTotalCost(0xc01c25f530, {0xc000a2bbd8, 0x1, 0xea63a0?})
        github.com/ethereum/go-ethereum/core/txpool/legacypool/list.go:497 +0xf0
github.com/ethereum/go-ethereum/core/txpool/legacypool.(*list).Remove(0xc01c25f530, 0xc002fd76e0)
        github.com/ethereum/go-ethereum/core/txpool/legacypool/list.go:445 +0x78
github.com/ethereum/go-ethereum/core/txpool/legacypool.(*LegacyPool).removeTx(0xc0004ff908, {0xef, 0x7, 0x96, 0xb6, 0x2f, 0xe0, 0xdd, 0x78, 0xe1, ...}, ...)
        github.com/ethereum/go-ethereum/core/txpool/legacypool/legacypool.go:1313 +0x2ec
github.com/ethereum/go-ethereum/core/txpool/legacypool.(*LegacyPool).loop(0xc0004ff908)
        github.com/ethereum/go-ethereum/core/txpool/legacypool/legacypool.go:429 +0x96b
created by github.com/ethereum/go-ethereum/core/txpool/legacypool.(*LegacyPool).Init in goroutine 1
        github.com/ethereum/go-ethereum/core/txpool/legacypool/legacypool.go:363 +0x429
````

--- a/core/txpool/legacypool/list.go
+++ b/core/txpool/legacypool/list.go
@@ -439,10 +439,14 @@ func (l *list) Cap(threshold int) types.Transactions {
 func (l *list) Remove(tx *types.Transaction) (bool, types.Transactions) {
 	// Remove the transaction from the set
 	nonce := tx.Nonce()
+	txInList := l.txs.Get(nonce)
+	if txInList == nil {
+		return false, nil
+	}
 	if removed := l.txs.Remove(nonce); !removed {
 		return false, nil
 	}
-	l.subTotalCost([]*types.Transaction{tx})
+	l.subTotalCost([]*types.Transaction{txInList})
 	// In strict mode, filter out non-executable transactions
 	if l.strict {
 		txs := l.txs.Filter(func(tx *types.Transaction) bool { return tx.Nonce() > nonce })

--- a/core/txpool/legacypool/list.go
+++ b/core/txpool/legacypool/list.go
@@ -443,6 +443,9 @@ func (l *list) Remove(tx *types.Transaction) (bool, types.Transactions) {
 	if txInList == nil {
 		return false, nil
 	}
+	if txInList.Hash().Cmp(tx.Hash()) != 0 {
+		return false, nil
+	}
 	if removed := l.txs.Remove(nonce); !removed {
 		return false, nil
 	}

